streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -2535,14 +2535,15 @@ abstract class ZStream[-R, +E, +O](val process: ZManaged[R, Nothing, ZIO[R, Opti
    */
   def toIterator: ZManaged[R, Nothing, Iterator[Either[E, O]]] =
     for {
-      pull    <- this.process.mapM(BufferedPull.make(_))
       runtime <- ZIO.runtime[R].toManaged_
+      pull    <- process
     } yield {
       new Iterator[Either[E, O]] {
+        var nextTake: Take[E, O] = null
```

Comment:
What do you think about storing
```
var chunkIterator: Iterator[O]
var error: E   
```
instead? 
This could save some CPU cycles on a random access to the chunk and on unwrapping `Take` multiple times/

Reply:
@simpadjo wouldn't this suggestion ignore failures though? or you mean `Iterator[Either[E, O]]`?

Reply:
@simpadjo this is what comes to mind:

```scala
new Iterator[Either[E, O]] {
  var nextTake: Exit[Option[E], Iterator[O]] = null

  def unsafeTake(): Unit =
    nextTake = runtime.unsafeRunSync(pull.map(_.toIterator))

  def hasNext: Boolean = {
    if (nextTake == null) {
      unsafeTake()
    }

    nextTake match {
      case Exit.Failure(cause) =>
        cause.failureOrCause match {
          case Left(None) => false
          case _          => true
        }
      case Exit.Success(it) => it.hasNext
    }
  }

  def next(): Either[E, O] = {
    if (nextTake == null) {
      unsafeTake()
    }

    nextTake match {
      case Exit.Failure(cause) =>
        cause.failureOrCause match {
          case Left(None)    => throw new NoSuchElementException("next on empty iterator")
          case Left(Some(e)) => Left(e)
          case Right(c)      => throw FiberFailure(c)
        }
      case Exit.Success(it) =>
        val value = it.next()

        if (!it.hasNext) {
          nextTake = null
        }

        Right(value)
    }
  }
}
```
Track of failures won't be lost.

Reply:
You could store it in a separate `var` but nevermind: seems that low-level manipulations can be avoided altogether

```
      def unfoldPull: Iterator[Either[E, O]] = {
        runtime.unsafeRunSync(pull) match {
          case Exit.Success(chunk) => chunk.iterator.map(Right(_)) ++ unfoldPull
          case Exit.Failure(cause) =>  cause.failureOrCause match {
            case Left(None)    => Iterator.empty
            case Left(Some(e)) => Iterator.single(Left(e))
            case Right(c)      => throw FiberFailure(c)
          }
        }
      }
```

Reply:
@simpadjo nice! just updated with your solution